src/lib/core/partialcubeop.ts

```diff
@@ -95,3 +95,138 @@ export function halfTurn(
       }
     : { face, row, column };
 }
+
+/**
+ * Moves the input CubeIndex to the specified Face while retaining the row and column coordinates.
+ */
+function changeFace({ row, column }: CubeIndex, face: Face): CubeIndex {
+  return {
+    face,
+    row,
+    column,
+  };
+}
+
+/**
+ * Computes the new CubeIndex to which the facelet occupying the input CubeIndex would be sent by
+ * an M slice. Only the U, F, D, and B faces are ever affected by M slices. Calling mSlice
+ * on a 3x3 cube with `layer == 2` has this effect (affected indices are highlighed using square
+ * brackets):
+ *
+ *                     {000} {001} [002]
+ *                     {010} {011} [012]
+ *                     {020} {021} [022]
+ *
+ * {100} {101} {102}   {200} {201} [202]   {300} {301} {302}   [400] {401} {402}
+ * {110} {111} {112}   {210} {211} [212]   {310} {311} {312}   [410] {411} {412}
+ * {120} {121} {122}   {220} {221} [222]   {320} {321} {322}   [420] {421} {422}
+ *
+ *                     {500} {501} [502]
+ *                     {510} {511} [512]
+ *                     {520} {521} [522]
+ *
+ *                                       |
+ *                                       v
+ *
+ *                     {000} {001} [420]
+ *                     {010} {011} [410]
+ *                     {020} {021} [400]
+ *
+ * {100} {101} {102}   {200} {201} [002]   {300} {301} {302}   [522] {401} {402}
+ * {110} {111} {112}   {210} {211} [012]   {310} {311} {312}   [512] {411} {412}
+ * {120} {121} {122}   {220} {221} [022]   {320} {321} {322}   [502] {421} {422}
+ *
+ *                     {500} {501} [202]
+ *                     {510} {511} [212]
+ *                     {520} {521} [222]
+ *
+ * @param {CubeIndex} position The starting position of the facelet to be moved
+ * @param {number} layer The column on the U face which should be affected by this operation
+ * @param {number} dimension The dimension of the cube being acted on
+ */
+export function mSlice(
```

Comment:
why is `mSlice` a different operation compared to say the standard R, L turns?

Reply:
So these aren't "real" cube operations, just partial ones that are meant to be utilities to aid in implementing actual cubeops. For instance, R on a 3x3 will be implemented using a combination of partial.quarterTurn on the R face and partial.reverseMSlice on layer 2. Hopefully that answers your other questions as well.